File: 2013 ACM-ICPC North American Qualifier/goldbach2.py
import math

def is_prime(n):
    if n == 1:
        return False
    for i in range(2, int(math.sqrt(n)) + 1):
        if n % i == 0:
            return False
    return True

def main():
    primes = [2]
    
    for i in range(3, 32000, 2):
        if is_prime(i):
            primes.append(i)
    
    cases = int(input())
    
    for zax in range(cases):
        count = 0
        str_result = ""
        num = int(input())
        
        for i in range(len(primes)):
            low = primes[i]
            high = num - primes[i]
            
            if high < low:
                break
            
            if high in primes:
                count += 1
                str_result += f"{low}+{high}\n"
        
        print(f"{num} has {count} representation(s)")
        print(str_result, end="")

if __name__ == "__main__":
    main()

File: 2013 ACM-ICPC North American Qualifier/goldbach2.cs
using System;
using System.Collections.Generic;

class Goldbach2 {
    static bool IsPrime(int n) {
        if (n == 1) return false;
        for (int i = 2; i < (int)(Math.Sqrt(n) + 1); i++)
            if (n % i == 0)
                return false;
        return true;
    }
    
    static void Main() {
        List<int> primes = new List<int>();
        primes.Add(2);
        
        for (int i = 3; i < 32000; i += 2)
            if (IsPrime(i))
                primes.Add(i);
        
        int cases = int.Parse(Console.ReadLine());
        
        for (int zax = 0; zax < cases; zax++) {
            int count = 0;
            string str = "";
            int num = int.Parse(Console.ReadLine());
            
            for (int i = 0; i < primes.Count; i++) {
                int low = primes[i];
                int high = num - primes[i];
                
                if (high < low)
                    break;
                
                if (primes.Contains(high)) {
                    count++;
                    str += low + "+" + high + "\n";
                }
            }
            
            Console.WriteLine(num + " has " + count + " representation(s)");
            Console.Write(str);
        }
    }
}